src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,191 @@
+* `le_min_degree_mul` -- the min-degree of a product is at least the sum of the min-degrees,
+* `max_degree_add_le` -- the max-degree of a sum is at most the sup of the max-degrees,
+* `le_min_degree_add` -- the min-degree of a sum is at least the inf of the min-degrees.
+
+## Implementation notes
+
+The current plan is to state and prove lemmas about `finset.support_sup` with a "generic"
+degree/weight function `D` from the grading Type `A` to a somewhat ordered Type `B`.
+
+Next, the general lemmas get specialized twice:
+* once for `max_degree` (essentially a simple application) and
+* once for `min_degree` (a simple application, via `order_dual`).
+These final lemmas are the ones that likely get used the most.  The generic lemmas about
+`finset.support.sup` may not be used directly much outside of this file.
+
+To see this in action, you can look at the triple
+`(sup_support_mul_le, max_degree_mul_le, le_min_degree_mul)`.
+-/
+
+variables {R A B ι : Type*} [semilattice_sup B] [order_bot B]
+
+namespace add_monoid_algebra
+open_locale classical big_operators
+
+section general_results_assuming_semilattice_sup
+
+section semiring
+
+variables [semiring R]
+
+section add_only
+
+lemma sup_support_add_le (D : A → B) (f g : add_monoid_algebra R A) :
+  (f + g).support.sup D ≤ (f.support.sup D) ⊔ (g.support.sup D) :=
+begin
+  refine (finset.sup_le (λ d ds, _)),
+  cases finset.mem_union.mp (finsupp.support_add ds) with df dg,
+  { exact (finset.le_sup df).trans le_sup_left },
+  { exact (finset.le_sup dg).trans le_sup_right }
+end
+
+variables [has_add A] [has_add B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+
+lemma sup_support_mul_le {D : A → B} (Dm : ∀ {a b}, D (a + b) ≤ D a + D b)
+  (f g : add_monoid_algebra R A) :
+  (f * g).support.sup D ≤ f.support.sup D + g.support.sup D :=
+begin
+  refine (finset.sup_le (λ d ds, _)),
+  obtain ⟨a, af, b, bg, rfl⟩ : ∃ a, a ∈ f.support ∧ ∃ b, b ∈ g.support ∧ d = a + b,
+  { simpa only [finset.mem_bUnion, finset.mem_singleton, exists_prop] using f.support_mul g ds },
+  refine Dm.trans (add_le_add _ _);
+  exact finset.le_sup ‹_›,
+end
+
+end add_only
+
+lemma sup_support_list_prod_le [add_monoid A] [add_monoid B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b) :
+  ∀ F : list (add_monoid_algebra R A),
+    F.prod.support.sup D ≤ (F.map (λ f : add_monoid_algebra R A, f.support.sup D)).sum
+| [] := begin
+    rw [list.map_nil, finset.sup_le_iff, list.prod_nil, list.sum_nil],
+    exact λ a ha, by rwa [finset.mem_singleton.mp (finsupp.support_single_subset ha)]
+  end
+| (f::fs) := begin
+    rw [list.prod_cons, list.map_cons, list.sum_cons],
+    exact (sup_support_mul_le Dm _ _).trans (add_le_add_left (sup_support_list_prod_le _) _)
+  end
+
+end semiring
+
+lemma sup_support_pow_le [comm_semiring R] [add_monoid A] [add_monoid B]
```

Comment:
```suggestion
lemma sup_support_pow_le [semiring R] [add_monoid A] [add_monoid B]
```
maybe this should go in the previous `semiring` section!

Reply:
Moved to `semiring`.